### PR TITLE
Add SnabbVMX manual

### PR DESCRIPTION
**Pull request by dpino**
Covers the following main topics:
- Introduction: What's SnabbVMX.
- Installation: How to run it.
- Configuration: Main configuration files (SnabbVMX, lwAFTR's conf and binding-table). Examples.
- User guide: SnabbVMX tools walk-through.
- Testing & troubleshooting: How to solve the main problems when running SnabbVMX and SnabbVMX main tests overview. Also includes how to write down new end-to-end tests for SnabbVMX.

### CODE REVIEW

**Comment by mwiget** [comment]
Nice! One comment though on the snabbvmx config, 'setup; should be 'settings' in 
https://github.com/Igalia/snabb/pull/515/files#diff-2e48ee534c9fdd7cd11ad65224290d02R26

It is not clear what the defaults are for ingress_drop_*. In fact, I just realize I used a wrong setting to enable flush by using (ingress_drop_action = "flush"), whereas I should use ingress_drop_monitor = 'flush'. Is it correct, that by default we do flush, so I can simply remove it when generating the snabbvmx config?

**Comment by dpino** [comment]
@mwiget I reviewed the ingress-drop-monitor related code. You're correct the settings attribute is named ingress_drop_monitor and not ingress_drop_action. I can rename it if you like it.

Then regarding default initialization value, what's happening right now is that without an ingress_drop_monitor value the timer is not initialized (it initializes to flush by default). So maybe what I am going to do is to initialize the timer is at least one ingress_drop monitor parameter is passed in settings (since ingress drop monitor takes default values for many other fields too). How does it sound?

**Comment by mwiget** [comment]
@dpino. I fixed the generation of the config already. Anyhow, the default setting is actually great (to flush by default), so I don't expect users actually changing this. 
I'm a bit confused about the timer. If default is to flush, then that means there is a periodic timer monitoring the ifInDiscards, correct? I actually see the flush() kick in without a config.

**Comment by dpino** [comment]
@mwiget Yes, that's correct. The timer periodically computes every `ingress_drop_interval` seconds the amount of ingress packet drops and calls `jit.flush` is action is 'flush', if not it simply prints out a warning message.

https://github.com/Igalia/snabb/blob/lwaftr/src/lib/timers/ingress_drop_monitor.lua#L74

I thought the timer in SnabbVMX was only run if it was set in the config file, but I realized there are default values for several ingress drop parameters in `lwaftr.lua`, and that's why it's running even if it's not set in the configuration file.

https://github.com/Igalia/snabb/blob/lwaftr/src/program/snabbvmx/lwaftr/lwaftr.lua#L95

Shall I get rid of these default values and start the timer only if it's set in the config file? As for those values that are not set the timer will be assigned default values.

https://github.com/Igalia/snabb/blob/lwaftr/src/lib/timers/ingress_drop_monitor.lua#L22

**Comment by mwiget** [comment]
Leave it as is. It is good to run it per default. If, for whatever reason, someone wants to just print out a warning, then we have a knob to do so.

File: src/program/snabbvmx/doc/README.configuration.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+command.
+
+It contains a reference to a Snabb's **lwAFTR** configuration file (which
+contains a reference to a binding-table).
+
+**snabbvmx-lwaftr-xe0.cfg**
+
+```lua
+return {
+   lwaftr = "snabbvmx-lwaftr-xe0.conf",
```
Comment (teknico) [inline src/program/snabbvmx/doc/README.configuration.md:16]: So the two config files have the same name, only differing in the extension? That could be confusing. Can we differentiate the names too?

File: src/program/snabbvmx/doc/README.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+## App network
+
+SnabbVMX's app network:
+
+![SnabbVMX](.images/snabbvmx.png)
+
+SnabbVMX works in one single 10Gb NIC.  Incoming traffic gets to an Intel82599 app
+managed by Snabb.
+
+App V4V6 splits traffic between IPv4 or IPv6 and forwards it to the correspondent
```
Comment (teknico) [inline src/program/snabbvmx/doc/README.md:28]: "correspondent" is a false friend (same as in Italian). You want something like "related" or "matching".

File: src/program/snabbvmx/doc/README.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        --sock /tmp/vhuser.sock
+```
+
+#### Configuration file
+
+`snabbvmx-lwaftr.cfg`
+
+```
+return {
+   lwaftr = "snabbvmx-lwaftr.conf",
```
Comment (teknico) [inline src/program/snabbvmx/doc/README.md:58]: Same remark as before: only the extension differs between the two filenames.

File: src/program/snabbvmx/doc/README.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+      vlan = false,
+   },
+}
+```
+
+lwaftr points to Snabb's lwAFTR configuration file.
+
+Othe attributes are further refined by SnabbVMX.  Attributes `ipv6_interface`
+and `ipv4_interface` are mandatory and ipv4_interface must include the IP
+addresses of the correspondent lwAFTR IPv4 interfaces. The IPv4 address is
```
Comment (teknico) [inline src/program/snabbvmx/doc/README.md:76]: "correspondent": see above.

File: src/program/snabbvmx/doc/README.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+SnabbVMX uses a different approach.  It setups a network design where only lw4o6
+packets are managed by the lwAFTR app, while everything else is forwarded to a VM
+which is in charge of providing ping, ARP and NDP resolution.
+
+## App network
+
+SnabbVMX's app network:
+
+![SnabbVMX](.images/snabbvmx.png)
```
Comment (teknico) [inline src/program/snabbvmx/doc/README.md:23]: Github is showing the image link as broken in the preview, even though it seems correct and the image file is there.

**Comment by dpino** [comment]
@mwiget I pushed an extra commit documenting ingress drop monitor default values. In addition, I added support for value 'off' in ingress_drop_monitor to deactivate the monitor, as Snabb's lwAFTR off does.

File: src/program/snabbvmx/doc/README.userguide.md [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+
+**CONF** is a SnabbVMX configuration file and **V4-IN** and **V6-IN** are the
+incoming V4 and V6 interfaces. The tools reads packets from a pcap file. **V4-OUT**
+and **V6-OUT** are the resulting packets after the lwAFTR processing. How the
+packets are processed depends on the configuration file and the binding table.
+
+Although SnabbVMX works in one single interface, snabbvmx check requires that
+the packet split is already done and provides an split output too.
+
+Imagine you have detected an error in the lwAFTR. First step is to obtain the
```
Comment (teknico) [inline src/program/snabbvmx/doc/README.userguide.md]: The text from here to the end of this chapter is the same as the **How to write a SnabbVMX end-to-end test** chapter at the end of `README.troubleshooting.md`. Is that intended?

**Review by teknico — CHANGES_REQUESTED** [review]
Only request for comments here, the minor changes are in a separate commit.

File: src/program/snabbvmx/doc/README.configuration.md (hunk 1 above)
Comment (dpino) [reply src/program/snabbvmx/doc/README.configuration.md:16]: I see your point but it's how it normally goes. For instance, the config files in tests/conf follow the same pattern. On the other hand, this pattern has the advantage that it makes it easier to know what's the correspondent Snabb's lwAFTR config file for a SnabbVMX configuration.

**Comment by teknico** [comment]
@dpino, Github is hiding some of my comments because of the commit I added afterwards. Could you please unhide them and reply anyway? Specifically the ones about the "correspondent" false friend (which is not used as an adjective in English), and the repeated section.

Also, please let me know if you happen to dislike any of the changes I made in my commit.

File: src/program/snabbvmx/doc/README.userguide.md (hunk 6 above)
Comment (dpino) [reply src/program/snabbvmx/doc/README.userguide.md]: No, it's a mistake. How to write a SnabbVMX end-to-end should be only in the troubleshooting guide.

File: src/program/snabbvmx/doc/README.configuration.md (hunk 1 above)
Comment (mwiget) [reply src/program/snabbvmx/doc/README.configuration.md:16]: This is an example of naming config files that I use in the vmx-docker-lwaftr container [https://github.com/mwiget/vmxlwaftr](url) and indeed chosen to easily match configs together. Anyone is free to chose other naming schema's based on their preference. There are no constraints imposed really.

Comment (dpino) [reply src/program/snabbvmx/doc/README.configuration.md:16]: If possible I prefer to keep the same schema. I think I get used to it :)

File: src/program/snabbvmx/doc/README.md (hunk 5 above)
Comment (dpino) [reply src/program/snabbvmx/doc/README.md:23]: It seems I needed to `git -f add` the image to actually add it. Now it's showing a link (the image is actually a symbolic link to an image in the parent folder).

File: src/program/snabbvmx/doc/README.configuration.md (hunk 1 above)
Comment (teknico) [reply src/program/snabbvmx/doc/README.configuration.md:16]: Sure, was just an idea. :-)

File: src/program/snabbvmx/doc/README.md (hunk 5 above)
Comment (teknico) [reply src/program/snabbvmx/doc/README.md:23]: Uhm, Github is still showing the image as missing though. Click on the **View** button on the top right to check it for yourself.

**Review by teknico — COMMENTED** [review]
I'm ok with the changes, once the link to the image is fixed and tests pass, even manually. :-)

**Review by teknico — APPROVED** [review]

**Comment by dpino** [comment]
Test passed. Here is "make test" result: https://gist.github.com/dpino/30d81ce6e7ac5c524f9057493c646aca

Regarding the image what I see when I click View is not the image but a string to the pointed image. I think it's correct.

**Comment by kbara** [comment]
'correspondent' was wrong, but 'corresponding' would be fine English, fwiw.

**Comment by teknico** [comment]
> 'correspondent' was wrong, but 'corresponding' would be fine English, fwiw.

@kbara, indeed!

**Pull request merged**